### PR TITLE
Juju 3.x Support

### LINKED ISSUE #545: Regression in libjuju 3 ; API changes in (at least) running commands on a unit
2022-08-10 21:04:26 [INFO] Image architecture set to amd64                                                                                                                                                                                                    2022-08-10 21:04:26 [WARNING] unknown facade EnvironUpgrader                                                                                                                                                                                                  2022-08-10 21:04:26 [WARNING] unexpected facade EnvironUpgrader found, unable to decipher version to use                                                                                                                                                      2022-08-10 21:04:30 [INFO] async -> sync. cleaning up pending tasks: len: 7                                                                                                                                                                                   Traceback (most recent call last):                                                                                                                                                                                                                              File "/home/ubuntu/charms/charm-ceph-fs/.tox/func-target/bin/functest-configure", line 8, in <module>                                                                                                                                                           sys.exit(main())                                                                                                                                                                                                                                            File "/home/ubuntu/charms/charm-ceph-fs/.tox/func-target/lib/python3.6/site-packages/zaza/charm_lifecycle/configure.py", line 97, in main                          
    configure(model_name, funcs)                                                                                                                                                                                                                              
  File "/home/ubuntu/charms/charm-ceph-fs/.tox/func-target/lib/python3.6/site-packages/zaza/charm_lifecycle/configure.py", line 58, in configure    
    run_configure_list(functions)
  File "/usr/lib/python3.6/contextlib.py", line 52, in inner                   
    return func(*args, **kwds)                                                                       
  File "/home/ubuntu/charms/charm-ceph-fs/.tox/func-target/lib/python3.6/site-packages/zaza/charm_lifecycle/configure.py", line 44, in run_configure_list    
    utils.get_class(func)()                                 
  File "/home/ubuntu/charms/charm-ceph-fs/.tox/func-target/lib/python3.6/site-packages/zaza/openstack/charm_tests/glance/setup.py", line 167, in add_lts_image    
    properties=properties)                                                                                                                
  File "/home/ubuntu/charms/charm-ceph-fs/.tox/func-target/lib/python3.6/site-packages/zaza/openstack/charm_tests/glance/setup.py", line 83, in add_image    
    glance_client = glance_client or _get_default_glance_client()                         
  File "/home/ubuntu/charms/charm-ceph-fs/.tox/func-target/lib/python3.6/site-packages/zaza/openstack/charm_tests/glance/setup.py", line 41, in _get_default_glance_client    
    keystone_session = openstack_utils.get_overcloud_keystone_session()
  File "/home/ubuntu/charms/charm-ceph-fs/.tox/func-target/lib/python3.6/site-packages/zaza/openstack/utilities/openstack.py", line 535, in get_overcloud_keystone_session    
    get_overcloud_auth(model_name=model_name),                                                                                 
  File "/home/ubuntu/charms/charm-ceph-fs/.tox/func-target/lib/python3.6/site-packages/zaza/openstack/utilities/openstack.py", line 2129, in get_overcloud_auth    
    model_name=model_name)                                                                                                                
  File "/home/ubuntu/charms/charm-ceph-fs/.tox/func-target/lib/python3.6/site-packages/zaza/utilities/juju.py", line 442, in leader_get    
    raise model.CommandRunFailed(cmd, result)                                             
zaza.model.CommandRunFailed: Command leader-get --format=yaml admin_passwd failed with code None, output None and error None

### PR DESCRIPTION
**Pull request by freyes**
This pull request adds juju-3.x support.

Summary of changes:
- Merged libjuju-3.1 branch
- Support for juju 3.2 and 3.3 support in the CI was not added due to issue #645

By default when installing 'zaza' won't have a pinned libjuju, hence users will get the latest juju available in pypi, although downstream consumers are encouraged to pin the library to the version compatible with the juju controller they want to connect to.

Running functional tests locally (`tox -e func-target -- first`) will use libjuju-2.9 by default, if contributors need/want to use a different version they can override the constraint setting the PIP_CONSTRAINTS environment variable, e.g.:

```
env PIP_CONSTRAINTS=$(pwd)/constraints-juju31.txt tox -e func-target --recreate -- first
```

Closes #545

### CODE REVIEW

**Comment by freyes** [comment]
:warning:  this is a breaking change, any thoughts on how we could enable a smoother transition are much appreciated.

**Comment by freyes** [comment]
Added a new commit that gives a more sane default https://github.com/openstack-charmers/zaza/pull/608/commits/084fcedffb3080fa4ee17d80b98f6d42fba025ee

```
 Depend on 'juju' when no extra is passed.

By default depend on 'juju' (no pinning), the pinning only comes into
place when an explicit extra is passed.

This makes the extras effectively nothing more than a pinning alias,
basically `pip install juju<3.0 zaza` becomes equivalent to `pip install
zaza[juju-29]`.
```

**Review by coreycb — COMMENTED** [review]
Hi Felipe, I added a couple of comments.

**Review by ajkavanagh — APPROVED** [review]
I'm generally okay, note @coreycb comments on using `_normalize_action_object`.

**Comment by rgildein** [comment]
Hi folks, can we merge this one along with #617? Thanks

File: zaza/model.py [hunk 1]
```diff
@@ -567,7 +570,12 @@ async def async_run_on_unit(unit_name, command, model_name=None, timeout=None):
     model = await get_model(model_name)
     unit = await async_get_unit_from_name(unit_name, model)
     action = await unit.run(command, timeout=timeout)
-    results = action.data.get('results')
+    await action.wait()
+    results = None
+    try:
```
[the remark this answers is not in the document]
Comment (freyes) [reply zaza/model.py]: ack, addressed by 10351db2e583eb624e9a440c81ad15084ca8e680

File: zaza/model.py [hunk 2]
```diff
@@ -593,7 +601,12 @@ async def async_run_on_leader(application_name, command, model_name=None,
         is_leader = await unit.is_leader_from_status()
         if is_leader:
             action = await unit.run(command, timeout=timeout)
-            results = action.data.get('results')
+            await action.wait()
+            results = None
+            try:
```
[the remark this answers is not in the document]
Comment (freyes) [reply zaza/model.py]: ack, addressed by 10351db2e583eb624e9a440c81ad15084ca8e680

File: zaza/model.py [hunk 3]
```diff
@@ -2184,7 +2231,15 @@ async def _check_for_file(model):
         for unit in units:
             try:
                 output = await unit.run('test -e "{}"; echo $?'.format(path))
-                contents = output.data.get('results')['Stdout']
+                await output.wait()
```
[the remark this answers is not in the document]
Comment (freyes) [reply zaza/model.py:2228]: ack, addressed by 10351db2e583eb624e9a440c81ad15084ca8e680

File: zaza/model.py [hunk 4]
```diff
@@ -1096,6 +1135,7 @@ async def async_run_action(unit_name, action_name, model_name=None,
     unit = await async_get_unit_from_name(unit_name, model)
     action_obj = await unit.run_action(action_name, **action_params)
     await action_obj.wait()
+    action_obj = _normalise_action_object(action_obj)
     if raise_on_failure and action_obj.status != 'completed':
```
[the remark this answers is not in the document]
Comment (freyes) [reply zaza/model.py]: addressed by 7a5dba6f0c6095e05aaa6f42066f60445e24d2ea

**Comment by freyes** [comment]
constraints seem to not be making effect 
```
func-target: install_deps> python -I -m pip install -r /home/runner/work/zaza/zaza/test-requirements.txt -c /home/runner/work/zaza/zaza/constraints-juju-29.txt
...
  Installed /home/runner/work/zaza/zaza/.tox/func-target/lib/python3.10/site-packages/zaza-0.0.2.dev1-py3.10.egg
  Processing dependencies for zaza==0.0.2.dev1
  Searching for juju==3.2.2
  Best match: juju 3.2.2
  Adding juju 3.2.2 to easy-install.pth file
  detected new path './zaza-0.0.2.dev1-py3.10.egg'
```

**Comment by chanchiwai-ray** [comment]
Hi, I am wondering if we sure also wrap `_normalise_action_results` to this [line](https://github.com/openstack-charmers/zaza/blob/efbf27737e4abd4566a7e2583e624c84a6630fde/zaza/model.py#L1099), as in 
```python
def _normalise_action_object(action_obj):
    try:
        # libjuju 3.x
        action_obj.data['status'] = action_obj._status
        action_obj.data['results'] = _normalise_action_results(action_obj.results) # wrap here
    except (AttributeError, KeyError):
        # libjuju 2.x format, no changes needed
        pass

    return action_obj
```

Because currently running actions with not return an action object with normalised `action.results`

**Comment by freyes** [comment]
Switching to draft until this issue gets sorted out.

```
2024-02-06 18:07:38.229111 | focal-medium |   "msg": "Ooops! Snap installation failed while executing 'sh -c \"/usr/bin/snap install --classic --channel 3.1/stable charmcraft\"', please examine logs and error output for more details.",
```

https://openstack-ci-reports.ubuntu.com/artifacts/f0b/908183/1/check/charm-build/f0b6c6e/job-output.txt
https://review.opendev.org/c/openstack/charm-keystone/+/908183

**Comment by freyes** [comment]
Related PR: https://github.com/openstack-charmers/zosci-config/pull/309

**Comment by freyes** [comment]
Issue found here in the CI runs - https://github.com/juju/python-libjuju/issues/1028

**Comment by freyes** [comment]
Here I'm attaching the logs of running `env TEST_ZAZA_BUG_LP1987332=on PIP_CONSTRAINTS=$(pwd)/constraints-juju31.txt tox -e func-target -- third 2>&1 | tee func-target.thirds.juju31.log` successfully.

```
2024-02-20 11:26:53 [INFO] Events:
  Configure zaza.charm_tests.noop.setup.basic_setup:
    Start: 1708439209.878124
    Finish: 1708439209.8789518
    Elapsed Time: 0.000827789306640625
    PCT Of Run Time: 1
  Deploy Bundle:
    Start: 1708438869.2582512
    Finish: 1708438878.273779
    Elapsed Time: 9.015527725219727
    PCT Of Run Time: 3
  Prepare Environment:
    Start: 1708438867.0046818
    Finish: 1708438869.258144
    Elapsed Time: 2.253462076187134
    PCT Of Run Time: 1
  Test zaza.charm_tests.libjuju.tests.RegressionTest:
    Start: 1708439209.8796163
    Finish: 1708439213.3465443
    Elapsed Time: 3.466928005218506
    PCT Of Run Time: 2
  Test zaza.charm_tests.noop.tests.NoopTest:
    Start: 1708439209.8790593
    Finish: 1708439209.8796015
    Elapsed Time: 0.0005421638488769531
    PCT Of Run Time: 1
  Wait for Deployment:
    Start: 1708438878.2737825
    Finish: 1708439209.8773825
    Elapsed Time: 331.603600025177
    PCT Of Run Time: 96
Metadata: {}
```

Full logs: [func-target.thirds.juju31.log](https://github.com/openstack-charmers/zaza/files/14345725/func-target.thirds.juju31.log)

**Review by ajkavanagh — APPROVED** [review]

**Pull request merged**